service/history/queues/scheduler_monitor.go

```diff
@@ -0,0 +1,214 @@
+	"go.temporal.io/server/common/util"
+)
+
+var (
+	_ tasks.Monitor[Executable] = (*schedulerMonitor)(nil)
+	_ tasks.Monitor[Executable] = (*noopSchedulerMonitor)(nil)
+)
+
+var noopScheduleMonitor = &noopSchedulerMonitor{}
+
+var defaultSchedulerMonitorOptions = schedulerMonitorOptions{
+	aggregationCount:    100,
+	aggregationDuration: time.Second * 30,
+}
+
+type (
+	schedulerMonitor struct {
+		taskChannelKeyFn  TaskChannelKeyFn
+		namespaceRegistry namespace.Registry
+		timeSource        clock.TimeSource
+		metricsHandler    metrics.MetricsHandler
+		options           schedulerMonitorOptions
+
+		status     int32
+		shutdownCh chan struct{}
+
+		sync.Mutex
+		scheduleStats map[TaskChannelKey]*scheduleStats
+	}
+
+	schedulerMonitorOptions struct {
+		aggregationCount    int
+		aggregationDuration time.Duration
+	}
+
+	scheduleStats struct {
+		lastStartTime time.Time
+		totalLatency  time.Duration
+		numStarted    int
+
+		taggedMetricsHandler metrics.MetricsHandler
+		lastEmissionTime     time.Time
+	}
+
+	noopSchedulerMonitor struct{}
+)
+
+func newSchedulerMonitor(
+	taskChannelKeyFn TaskChannelKeyFn,
+	namespaceRegistry namespace.Registry,
+	timeSource clock.TimeSource,
+	metricsHandler metrics.MetricsHandler,
+	options schedulerMonitorOptions,
+) *schedulerMonitor {
+	return &schedulerMonitor{
+		taskChannelKeyFn:  taskChannelKeyFn,
+		namespaceRegistry: namespaceRegistry,
+		timeSource:        timeSource,
+		metricsHandler:    metricsHandler,
+		options:           options,
+
+		status:     common.DaemonStatusInitialized,
+		shutdownCh: make(chan struct{}),
+
+		scheduleStats: make(map[TaskChannelKey]*scheduleStats),
+	}
+}
+
+func (m *schedulerMonitor) Start() {
+	if !atomic.CompareAndSwapInt32(&m.status, common.DaemonStatusInitialized, common.DaemonStatusStarted) {
+		return
+	}
+
+	go m.metricEmissionLoop()
+}
+
+func (m *schedulerMonitor) Stop() {
+	if !atomic.CompareAndSwapInt32(&m.status, common.DaemonStatusStarted, common.DaemonStatusStopped) {
+		return
+	}
+
+	close(m.shutdownCh)
+}
+
+func (m *schedulerMonitor) RecordStart(executable Executable) {
+	startTime := m.timeSource.Now()
+	taskChanKey := m.taskChannelKeyFn(executable)
+
+	m.Lock()
+	defer m.Unlock()
+
+	stats := m.getOrCreateScheduleStatsLocked(taskChanKey)
+
+	latency := util.Min(
+		startTime.Sub(stats.lastStartTime),
+		startTime.Sub(executable.GetScheduledTime()),
+	)
+	stats.lastStartTime = startTime
+	stats.numStarted++
+	stats.totalLatency += latency
+
+	if stats.numStarted >= m.options.aggregationCount {
+		m.emitMetric(taskChanKey, stats)
+	}
+}
+
+func (m *schedulerMonitor) metricEmissionLoop() {
+	emissionTicker := time.NewTicker(m.options.aggregationDuration)
+	defer emissionTicker.Stop()
+
+	for {
+		select {
+		case <-m.shutdownCh:
+			return
+		case <-emissionTicker.C:
+			m.Lock()
+
+			now := m.timeSource.Now()
+			for taskChanKey, stats := range m.scheduleStats {
+				if now.Sub(stats.lastEmissionTime) > m.options.aggregationDuration {
```

Comment:
Some similarity with https://github.com/temporalio/temporal/pull/2970.
Maybe we could abstract some generic logic here as well (not in this PR).